File: assignment 3/problem3.py
import numpy as np
import matplotlib.pyplot as plt
from numpy.linalg import eig
from scipy.io import loadmat

if __name__=='__main__':
    data = loadmat('teapots.mat')
    X = data['teapotImages']
    u = np.mean(X, axis=0).reshape(1,1900)
    x = X - np.repeat(u,100,axis=0)
    C = x.T.dot(x)/(x.shape[0]-1)
    D, V = eig(C)
    print(x.shape)
    print(C)
    print(D[:3])
    print(V[:,:3])
    

File: assignment 3/problem3.m
load('teapots.mat')
data = teapotImages;
u = mean(data);
X = data - u;
C = cov(X);
[V, D] = eig(C);
[d, ind] = sort(diag(D),'descend');
d = d(1:3,:);
v = V(:,ind(1:3));
c = X*v;
X_hat = u+c*v';

for i = 1:10
    figure(i);
    colormap gray;
    subplot(1,2,1);
    imagesc(reshape(data(i,:),38,50));
    title('Original');
    axis image;
    subplot(1,2,2)
    imagesc(reshape(X_hat(i,:),38,50));
    title('PCA');
    axis image;
end

norm(data-X_hat)

